src/Microsoft.ML.Tokenizers/Model/BPE.cs

```diff
@@ -327,80 +317,124 @@ internal string CharToString(char c)
+                        length = 2;
+                        s = w.Slice(i, 2);
+                    }
+                    else
+                    {
+                        length = 1;
+                        s = w.Slice(i, 1);
+                    }
 
-                if (_vocab.TryGetValue(s, out int id))
-                {
-                    if (unk.HasValue)
+                    // Add the `continuing_subword_prefix` if relevant
+                    if (i > 0 && ContinuingSubwordPrefix is not null)
                     {
-                        word.Add(unk.Value.Id, unk.Value.Len);
-                        unk = null;
+                        if (buffer is null)
+                        {
+                            // 60 should be enough for most cases
+                            buffer = ArrayPool<char>.Shared.Rent(60);
+                        }
+
+                        if (ContinuingSubwordPrefix.Length + s.Length <= buffer.Length)
+                        {
+                            ContinuingSubwordPrefix.AsSpan().CopyTo(buffer.AsSpan());
+                            s.CopyTo(buffer.AsSpan().Slice(ContinuingSubwordPrefix.Length));
+                            s = buffer.AsSpan().Slice(0, ContinuingSubwordPrefix.Length + s.Length);
+                        }
+                        else
+                        {
+                            string s1 = s.Length == 1 ? CharToString(s[0]) : s.ToString();
+                            s = $"{ContinuingSubwordPrefix}{s1}".AsSpan();
+                        }
                     }
-                    word.Add(id, length);
-                }
-                else if (UnknownToken is not null)
-                {
-                    if (unk.HasValue)
+
+                    // Add the `end_of_word_suffix` if relevant
+                    if (i + length >= w.Length && EndOfWordSuffix is not null)
                     {
-                        if (FuseUnknownTokens)
+                        if (buffer is null)
                         {
-                            // Fuse unk
-                            unk = (unk.Value.Id, unk.Value.Len + length);
+                            // 60 should be enough for most cases
+                            buffer = ArrayPool<char>.Shared.Rent(60);
+                        }
+
+                        if (s.Length + EndOfWordSuffix.Length <= buffer.Length)
+                        {
+                            s.CopyTo(buffer.AsSpan());
+                            EndOfWordSuffix.AsSpan().CopyTo(buffer.AsSpan().Slice(s.Length));
+                            s = buffer.AsSpan().Slice(0, s.Length + EndOfWordSuffix.Length);
                         }
                         else
                         {
-                            // Do not fuse unk, add the previous one
+                            string s1 = s.Length == 1 ? CharToString(s[0]) : s.ToString();
+                            s = $"{s1}{EndOfWordSuffix}".AsSpan();
```

Comment:
On net6.0+, this can just be `s = $"{s}{EndOfWordSuffix}";` or `s = string.Concat(s, EndOfWordSuffix);`.